src/lib.rs

```diff
@@ -101,12 +124,38 @@ impl<T> Gc<T> {
 
             AllocMetadata::insert(objptr as usize, layout.size(), true);
             let headerptr = objptr.sub(size_of::<usize>());
-            ptr::write(headerptr as *mut usize, 1);
+            ptr::write(headerptr as *mut GcHeader, GcHeader::new());
             objptr as *mut T
         }
     }
 }
 
+impl GcHeader {
+    pub(crate) fn new() -> Self {
+        Self(0)
+    }
+
+    pub(crate) fn mark_bit(&self) -> bool {
+        ((self.0 >> 0) & 1) == 1
+    }
+
+    pub(crate) fn set_mark_bit(&mut self, mark: bool) {
+        if mark {
+            self.0 |= 1 << 0
+        } else {
+            self.0 &= !(1 << 0)
+        }
+    }
+}
+
+impl<T> Deref for Gc<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        unsafe { &*(self.objptr as *const T) }
```

Comment:
What's unsafe here? `&*`?

Reply:
Yes, it's de-referencing a raw pointer.

Reply:
Did you try without? It may not actually deref.

Reply:
In Rust you have to compose both operators if you want to get a reference from a raw pointer. It doesn't deref now, but the references are always considered dereference-able, so if you construct one from a raw pointer you better be sure that it's safe to de-reference it in the future.

Reply:
Right, but that didn't answer my question. Did you try it without?

(You are probably right)

Reply:
I'm not sure what you mean? `&(self.objptr as *const T)` is invalid Rust code.

Reply:
Turning a raw pointer into a Rust reference is always unsafe.

Reply:
> Turning a raw pointer into a Rust reference is always unsafe.

ok

Reply:
@vext01 Sorry, I understand now that you meant try and remove the unsafe block, not the `*`.